fix(api): surface subscription deletion errors to users

Mairuis opened the pull request:
## Summary

This PR fixes a bug where trigger subscription deletion/unsubscribe errors were silently ignored during the rebuild process.

- Check `UnsubscribeResult.success` and propagate error messages to users when unsubscribe fails
- Simplify rebuild logic by removing unnecessary try/except wrapper around unsubscribe
- Refactor update API to use cleaner conditional logic with `any()`
- Remove redundant test cases that validated silent error handling behavior

Fixes #30332

## Test Plan

- [x] Verify that unsubscribe errors are now properly surfaced to users with clear error messages
- [x] Verify that successful subscription rebuilds continue to work as expected
- [x] Run existing unit tests to ensure no regressions

Mairuis committed:
fix(api): surface subscription deletion errors to users

Previously, when rebuilding a trigger subscription, errors from the
unsubscribe operation were silently caught and logged without
propagating to the user. This left users unaware of failures during
subscription management.

Changes:
- Check UnsubscribeResult.success and raise ValueError with the error
  message when unsubscribe fails
- Simplify the rebuild logic by removing unnecessary try/except wrapper
- Refactor update API to use cleaner conditional logic
- Remove redundant test cases that tested silent error handling

Yeuoly commented on api/services/trigger/trigger_provider_service.py:
```diff
@@ -868,111 +868,50 @@ def rebuild_trigger_subscription(
-                        subscription_id=subscription.id,
-                    )
+        credential_type = CredentialType.of(subscription.credential_type)
+        if credential_type not in [CredentialType.OAUTH2, CredentialType.API_KEY]:
```
I suggest to use `set` rather than `list`

Yeuoly commented on api/services/trigger/trigger_provider_service.py:
```diff
@@ -868,111 +868,50 @@ def rebuild_trigger_subscription(
-                        provider_id=subscription.provider_id,
-                        subscription_id=subscription.id,
-                    )
+        credential_type = CredentialType.of(subscription.credential_type)
```
There is no need to use `CredentialType.of` as `CredentialType` is a kind of `StrEnum`

Mairuis replied:
I tried, but there are some CI errors 😭
<img width="1704" height="205" alt="image" src="https://github.com/user-attachments/assets/3ec58d08-7004-4da9-be66-c253887d92c4" />

Mairuis committed:
feat(api): add validation to ensure at least one field is provided in TriggerSubscriptionUpdateRequest

- Introduced a model validator in TriggerSubscriptionUpdateRequest to enforce that at least one of the fields (name, credentials, parameters, properties) must be provided.
- Refactored the TriggerSubscriptionUpdateApi to use the validated request object and simplified the logic for updating subscriptions based on the credential type.
- Updated the credential type check in TriggerProviderService to use a set for better performance and clarity.

Mairuis committed:
doc(api): enhance comments for clarity in TriggerSubscriptionUpdateApi and TriggerProviderService

Mairuis committed:
doc(api): improve comments for better understanding in TriggerSubscriptionUpdateApi and TriggerProviderService

Mairuis committed:
fix(api): fix basedpyright checks

Mairuis committed:
refactor(api): remove unsupported credential type test from TriggerProviderService

- Deleted the test for rebuilding trigger subscriptions with unsupported credential types, as it was deemed unnecessary.
- This change helps streamline the test suite by focusing on relevant scenarios.

Mairuis committed:
Merge branch 'main' into fix/surface-subscription-deletion-errors

Mairuis committed:
Merge branch 'main' into fix/surface-subscription-deletion-errors

Mairuis committed:
Merge branch 'main' into fix/surface-subscription-deletion-errors

Mairuis committed:
Merge branch 'main' into fix/surface-subscription-deletion-errors

Mairuis committed:
Merge branch 'main' into fix/surface-subscription-deletion-errors
